tests/cpp17/stringify_util.h

```diff
@@ -0,0 +1,195 @@
+  for (auto it = vec.cbegin(), end = vec.cend(); it != end; ++it) {
+    text += prologue;
+    text += StringifyFlatbufferValue(*it).value_or("(field absent)");
+    text += epilogue;
+  }
+  if (vec.cbegin() != vec.cend()) {
+    text.resize(text.size() - epilogue.size());
+  }
+  text += std::string("\n") + indent + "]";
+  return text;
+}
+
+template<typename T> std::string StringifyArithmeticType(T val) {
+  std::ostringstream oss;
+  oss << val;
+  return oss.str();
+}
+
+}  // namespace detail
+
+/*******************************************************************************
+** Take any flatbuffer type (table, struct, Vector, int...) and stringify it.
+*******************************************************************************/
+template<typename T>
+std::optional<std::string> StringifyFlatbufferValue(T &&val,
+                                                    const std::string &indent) {
+  constexpr bool is_pointer = std::is_pointer_v<std::remove_reference_t<T>>;
+  if constexpr (is_pointer) {
+    if (val == nullptr) {
+      return std::nullopt;  // Field is absent.
+    }
+  }
+  using decayed =
+      std::decay_t<std::remove_pointer_t<std::remove_reference_t<T>>>;
+
+  // Is it a Flatbuffers Table or Struct?
+  if constexpr (detail::is_fb_table_or_struct_v<decayed>) {
+    // We have a nested table or struct; use recursion!
+    if constexpr (is_pointer) {
+      return detail::StringifyTableOrStruct(*val, indent);
+    } else {
+      return detail::StringifyTableOrStruct(val, indent);
+    }
+  }
+
+  // Is it an 8-bit number?  If so, print it like an int (not char).
+  else if constexpr (std::is_same_v<decayed, int8_t> ||
+                     std::is_same_v<decayed, uint8_t>) {
+    return detail::StringifyArithmeticType(static_cast<int>(val));
+  }
+
+  // Is it an enum? If so, print it like an int, since Flatbuffers doesn't yet
+  // have type-based reflection for enums, so we can't print the enum's name :(
+  else if constexpr (std::is_enum_v<decayed>) {
+    return StringifyFlatbufferValue(
+        static_cast<std::underlying_type_t<decayed>>(val), indent);
+  }
+
+  // Is it an int, double, float, uint32_t, etc.?
+  else if constexpr (std::is_arithmetic_v<decayed>) {
+    return detail::StringifyArithmeticType(val);
+  }
+
+  // Is it a Flatbuffers string?
+  else if constexpr (std::is_same_v<decayed, flatbuffers::String>) {
+    return "\"" + val->str() + "\"";
+  }
+
+  // Is it a Flatbuffers Vector?
+  else if constexpr (detail::is_fb_vector_v<decayed>) {
+    return detail::StringifyVector(*val, indent);
+  }
+
+  // Is it a void pointer?
+  else if constexpr (std::is_same_v<decayed, void>) {
+    // Can't format it.
+    return std::nullopt;
+  }
+
+  else {
+    // Not sure how to format this type, whatever it is.
+    static_assert(sizeof(T) != sizeof(T),
+                  "Do not know how to format this type T (the compiler error "
+                  "should tell you nearby what T is).");
+  }
```

Comment:
Note to reviewers: I changed it so that if it gets to the bottom of the function and can't format a type, it generates a compile error instead of a runtime error, since that is really a compile-time thing.  Before it was just returning `std::nullopt`.  This means that if you add new types to `Monster` that are not covered by the cases above then it will notify you with a compile error instead of silently omitting the field, which it was doing before.

So now, this function only returns `std::nullopt` when 1) it knows how to format a types, and 2) that type is "absent", e.g. a table/struct's pointer is null.